Make Data::Dumper strict and warnings compliant

toddr opened the pull request.

toddr committed:
Make Data::Dumper strict and warnings compliant

atoomic commented on dist/Data-Dumper/Dumper.pm:
```diff
@@ -9,9 +9,12 @@
 BEGIN {
-    $VERSION = '2.175'; # Don't forget to set version and release
-}               # date in POD below!
+    our $VERSION = '2.176'; # Don't forget to set version and release
```
I still believe this is weird to define these variables in a `BEGIN` block... they should be defined before and set inside the block..

atoomic replied:
```perl
use strict;
use warnings;
use v5.30;
BEGIN {
    our $X = 42;
}
say $X;
```
```
> perl t.pl
Variable "$X" is not imported at t.pl line 10.
Global symbol "$X" requires explicit package name (did you forget to declare "my $X"?) at t.pl line 10.
Execution of t.pl aborted due to compilation errors.
```

Grinnz replied:
It is not weird at all. The variable is not needed after the BEGIN block.

atoomic commented on dist/Data-Dumper/Dumper.pm:
```diff
@@ -23,9 +26,10 @@ use constant IS_PRE_516_PERL => $] < 5.016;
+    our @EXPORT = qw(Dumper);
+    our @EXPORT_OK = qw(DumperX);
+    our $Useperl;
 
```
same for these variables ^

Grinnz replied:
The only variable here used outside this scope is `$Useperl`, but since it is aliased again with `our`, it will refer to the same package variable already.

Grinnz replied:
This `our` declaration can actually be done inline in the assignment below, thus not needing this extra line.

atoomic commented on dist/Data-Dumper/t/dumper.t:
```diff
@@ -93,26 +103,36 @@ sub convert_to_native($) {
     return join "", grep { defined } @output;
 }
 
+sub add_my_to_dump {
```
this helper seems to be used in more than one location, maybe it could be shared

toddr replied:
I agree but I'm only adding it here for this work and it is in quite a few tests already. I'd prefer to take it out in a follow up commit.

Leont commented on dist/Data-Dumper/t/qr.t:
```diff
@@ -16,9 +19,15 @@ use Data::Dumper;
     ok(!$@, "Dumping $qr with PP") or diag $@, Dumper $qr;
 }
+
+sub add_my_to_dump {
```
This makes the test pass, but is it really the right thing to do?

It means that any code that assumes `eval Dumper $foo` will work will break, and I think there's quite a bit of such code out there.

Leont reviewed (changes requested):
This makes the test pass, but I don't think that makes it the right thing to do.

It means that any code that assumes «eval Dumper $foo» will work will break, and I suspect there's quite a bit of such code out there. I don't think that's an acceptable outcome.

tonycoz commented on dist/Data-Dumper/t/bless_var_method.t:
```diff
@@ -1,9 +1,13 @@
+use warnings;
+
+use Config;
+
 BEGIN {
     if ($ENV{PERL_CORE}){
-        require Config; import Config;
         no warnings 'once';
         if ($Config{'extensions'} !~ /\bData\/Dumper\b/) {
             print "1..0 # Skip: Data::Dumper was not built\n";
```
This block could probably just be removed.  TEST (and hence harness) already skip the tests.

toddr replied:
cool. I'll add a commit

toddr committed:
Remove unnecessary need for Data::Dumper Config{extensions} check.

These unit test checks are unnecessary since the entire module
isn't tested if Data::Dumper is excluded.

toddr committed:
Remove the need for add_my_to_dump helper in Data::Dumper.

Leont commented:
> > It means that any code that assumes «eval Dumper $foo» will work will break, and I suspect there's quite a bit of such code out there. I don't think that's an acceptable outcome.
> 
> I'm not sure what line you're commenting on.

All tests that use `add_my_to_dump` are wrong. Those tests are currently testing `eval Dumper $foo`, and that's exactly what they should be doing.

toddr commented:
> All tests that use `add_my_to_dump` are wrong. Those tests are currently testing `eval Dumper $foo`, and that's exactly what they should be doing.

It was always the case that code which did `use strict` needed to first declare `my $VAR1` or `no strict 'vars'` at a minimum in order for `eval Dumper $foo` to work.

wchristian commented:
The problem is that:

1. This must work: `my $blah = eval Dumper $foo;`
2. Regardless of what else is decided, the tests for Data::Dumper must continue to verify that the above still works for Perls that aren't strict-by-default.

toddr commented:
> Regardless of what else is decided, the tests for Data::Dumper must continue to verify that the above still works for Perls that aren't strict-by-default.

None of these changes are for a strict perl. See updated commits. IMO this is just the same fix in another way. I'm not convinced what you guys are worried about is a big deal or unknown.

Grinnz commented:
You have changed what is being tested. If anything, `no strict 'vars'` can be declared when testing the eval since that's actually what's required.

wchristian commented:
@leonerd doesn't have the time right now, but he has requested to be given some time until he can comment.

toddr commented:
> You have changed what is being tested. If anything, `no strict 'vars'` can be declared when testing the eval since that's actually what's required.

how is `no strict 'vars'` any better than `my $VAR1` ?

Grinnz commented:
Declaring `my $VAR1` changes the test, you are no longer testing that the string eval will execute and have a certain result without that declaration in scope (since obviously, you added it because it doesn't work without that in scope). Whereas declaring `no strict 'vars'` changes the hints to what is required for string-eval of Dumper output to function, as it tested before.

toddr commented:
> Declaring `my $VAR1` changes the test

No argument here. So I'm clear, your concern is that testing `eval Dumper $foo` WITH strict is somehow less useful and one day that is going to break with strict off and we're not going to notice?

Grinnz commented:
There are many ways the changed test could break that have no bearing on existing usage of string-eval Dumper output. In the end it is testing/guaranteeing behavior that does not function with strict 'vars', so that's how it should be tested.

toddr commented:
> The problem is that:
> 1. This must work: `my $blah = eval Dumper $foo;`

I just came back and looked at this. The problem here is the this doesn't actually work:

```
$>perl5.30.0 -w -e'use strict; use Data::Dumper; my $foo = { a => 5, b => 7 }; my $d = eval Dumper $foo; print $@'       
Global symbol "$VAR1" requires explicit package name (did you forget to declare "my $VAR1"?) at (eval 5) line 1.
```

Grinnz commented:
It does, without strict vars, as it has forever.

toddr commented:
> > I'm not sure what line you're commenting on.
> 
> All tests that use `add_my_to_dump` are wrong. Those tests are currently testing `eval Dumper $foo`, and that's exactly what they should be doing.

Does this meet your concern abd02b554586427dff88409c0fb529269f8f181b ? I'll point out that some of this code was already the way you indicated was wrong.

haarg commented:
`dumper.t` is still using `add_my_to_dump`.

toddr commented:
> `dumper.t` is still using `add_my_to_dump`.

As it has for years. I consider it out of scope for this change. If you would like to submit a pull request I would appreciate it!￼

toddr commented:
> `dumper.t` is still using `add_my_to_dump`.

You're right. This was even the one that `add_my_to_dump` got injected into. Sorry there's been so many rebases that I missed this one. This turned out to the the most complex one to fix. I think I've got it now.

toddr commented:
I believe I have addressed all of the concerns that have been raised. The only possible question is if the tests need further refinement. I've not heard back from any of the people that were concerned. To my knowledge, there are no concerns about the module itself?

Barring any further concerns, I'm going to merge this in 48 hours.

karenetheridge commented:
> Barring any further concerns, I'm going to merge this in 48 hours.

Please don't issue ultimatums. A tight timeline for non-critical PRs is not reasonable.

xsawyerx commented:
> > Barring any further concerns, I'm going to merge this in 48 hours.
> 
> Please don't issue ultimatums. A tight timeline for non-critical PRs is not reasonable.

Incorrect.

It is a common pattern in our issues (prior to GitHub) in which we provide a clear deadline for progress on the ticket. Todd is merely using this mechaism, also used abundantly by Jim K.

In fact, I think this helps us avoid warnocking and languishing on tickets, which we are quite familiar with and want to avoid doing.

Luckily, we are using Git and can - if necessary - revert. We've done this sucessfully in the past and can do it again.

xsawyerx reviewed (approved):
The primary concerns were that the test for `Data::Dumper` in a certain situation had to be reinstated. It was addressed. This change does not break `Data::Dumper` itself or the code using it.

With Tony's additional blessing, I see no reason to postpone merging this. If one comes up, you can always re-open the ticket.

karenetheridge commented:
*puts on moderator hat*

@xsawyerx I received private communications about this PR expressing concerns with the tone therein. "I'm going to merge this in 48 hours" is unnecessarily hostile, and given the current situation we could do with a lot less hostility.  "If there are no more concerns, I'd like to merge this soon" would have been much better.

Everyone, please remember to be kind. It's a difficult year for everyone, and some people are also bearing burdens that we are not aware of.

hvds commented:
> Everyone, please remember to be kind.

Yes, and beyond that please everyone Assume Good Faith. While I can see just about imagine how someone could read hostility or an ultimatum out of "Barring any further concerns, I'm going to merge this in 48 hours", an assumption of good faith would surely quickly rule those out in favour of seeing it as a simple attempt to keep things moving forward.

xsawyerx commented:
> _puts on moderator hat_
> 
> @xsawyerx I received private communications about this PR expressing concerns with the tone therein. "I'm going to merge this in 48 hours" is unnecessarily hostile,

I do not consider this hostile. We've had numerous responses to uncontentious tickets that included comments like "with no further comments, I intend to close this ticket" or "I will merge this within a few days." This is *far* from uncommon for our tickets. This ticket had feedback, it was reviewed and addressed. I do not see any hostility in a comment like "I'm going to merge this in 48 hours" after an additional period.

In a recent PR, I wrote something similiar. No one raised alarms, concerns, the author did not feel this was hostile in any way. I think this ticket was misunderstood and thus tensions might be high, but stepipng back, the comment Todd made had nothing out of the ordinary in it.

> and given the current situation we could do with a lot less hostility. "If there are no more concerns, I'd like to merge this soon" would have been much better.

I agree with you on both of these sentences.

To that end, I think it's worth mentioning that a response of "Please don't issue ultimatums" suggests an intention by Todd stemming from a rather routine ticket comment ("I'm going to merge this within X time", or as we write in other tickets "I'm going to close this in X time").

A comment of "I suggest we keep this ticket open for Y time to allow people additional feedback" would have been better, expressing the need (keeping it open for comments) rather than just reviewing this ordinary comment by Todd as an "ultimatum."

I did look up ultimatum, since English is not my primary language. Definition: "a final demand or statement of terms, the rejection of which will result in retaliation or a breakdown in relations."

> Everyone, please remember to be kind. It's a difficult year for everyone, and some people are also bearing burdens that we are not aware of.

Agreed.

wchristian commented:
> I think this helps us avoid warnocking and languishing on tickets, which we are quite familiar with and want to avoid doing.

This form of setting a deadline is a very unilateral tool that is not equally available. It may help reduce warnocking, but it does so unevenly in a very non-systemic manner. Language as was used above will be felt differently by people in different positions in these power equations.

> In a recent PR, I wrote something similiar. No one raised alarms, concerns, the author did not feel this was hostile in any way.

I feel there could be a significant difference between those two situations.

I don't know if this happened in your ticket, but in this one, the timeline went:

--------------------------

> request made: leonerd doesn't have the time right now, but he has requested to be given some time until he can comment.

*leonerd doesn't yet find the time amidst EVERYTHING going on in the world and in perl right now*

> declaration made: Barring any further concerns, I'm going to merge this in 48 hours.

--------------------------

This order of event was interpreted rather harshly, and it's totally valid to say "i didn't mean it that harshly", but i don't think the interpretation was invalid.

wchristian commented:
Leonerd managed to spend a little time on this, and his judgement is:

> 22:57 <LeoNerd> Then this is all fine

Which clears up the last doubt on this. :)

leonerd commented:
+1 to what @wchristian . Having reviewed the current set of changes, the `no strict 'refs';` appears to be the right solution. So I'm happy.

xsawyerx commented:
> > I think this helps us avoid warnocking and languishing on tickets, which we are quite familiar with and want to avoid doing.
> 
> This form of setting a deadline is a very unilateral tool that is not equally available. It may help reduce warnocking, but it does so unevenly in a very non-systemic manner. Language as was used above will be felt differently by people in different positions in these power equations.

I don't entirely understand. This comment offered additional time and stated that - unless additional comments within this additional time will be given - the next action will be performed on this. It explicitly provided additional time *and* offered that it will only proceed if no additional comments are given. This was also given after three days with no comment by anyone with all issues resolved, as far as Todd was aware. (Except Paul - see below.)

To note this as an unequal tool is irrelevant to the fact bercause:

1. Obivously, only people with merge capabilities could make this. People who have no merge capabilities cannot use this tool. (This is equivalent to saying that a shopkeep who wishes to close their store at 10 pm should not make this call because it's unequal - they have the keys to the store but the patrons do not.)

2. This is a tool frequently used both in Open Source and Free Software as well as in Perl. I wrote a similar comment yesterday. Jim writes many such comments. We say this about features, let alone such small changes as making a module strict-compatible. Why were none of those "hostile" or "unfair"?

Considering everything was resolved, no additional comments made in three days, in other situations we might have just merged it and moved on. Todd made sure people on this thread are aware that he intends to move forward and offered an additional period to provide comments.

The comment "Hey, can you keep this open for another week?" would have worked to keep it open. "This issue wasn't resolved yet" would have kept it open.

Three days with all issues resolved and no comment, another request for comment, and two additional days - I don't see how anything he did is wrong here.

This wasn't even a contentious change. It was misunderstood as breaking code but it only rendered a particular test irrelevant and that test had to be reintroduced in a relevant manner.

> > In a recent PR, I wrote something similiar. No one raised alarms, concerns, the author did not feel this was hostile in any way.
> 
> I feel there could be a significant difference between those two situations.
> 
> I don't know if this happened in your ticket, but in this one, the timeline went:
> 
> > request made: leonerd doesn't have the time right now, but he has requested to be given some time until he can comment.
> 
> _leonerd doesn't yet find the time amidst EVERYTHING going on in the world and in perl right now_

Todd reached out to Paul and spoke to him. Todd also added Tony as a reviewer. Todd is still following-up on this after Paul intends to have another look at it - that's what I understood. This doesn't mean that Todd cannot comment here giving a notification for others while he's following up with Paul on his concerns.

FWIW, a comment of "Have you addressed Paul's concerns?" or "Wait, Paul hasn't replied yet" would have been more than enough.

> > declaration made: Barring any further concerns, I'm going to merge this in 48 hours.
> 
> This order of event was interpreted rather harshly, and it's totally valid to say "i didn't mean it that harshly", but i don't think the interpretation was invalid.

I think the missing context here of Todd reaching out directly to Paul to further discuss it directly is what allowed for this interpretation. I suggest we all take a note of @hvds' book and be prudent in our judgment of each other.

wchristian commented:
> I don't entirely understand.

The important part was this. :)

> Language as was used above will be felt differently by people in different positions in these power equations.

That means, regardless of how correct the facts of a situation was, without the full extent of all the facts being known to all parties people will have different feelings on matters. That's just how humans are.

While facts do indeed not care about feelings, it is also a fact that feelings exist. :)

As such, i think it's equally important that people correcting themselves like "i didn't mean it like that, i meant xyz" be accepted; but also that people not be blamed for believing initially that words were meant as they were written.

And if Paul was contacted, then yay. Nobody knew. If anybody had known, the words would've been perceived differently.

At this point most anyone i'm aware of is just glad that things are resolved. :)

xsawyerx commented:
> > I don't entirely understand.
> 
> The important part was this. :)
> 
> > Language as was used above will be felt differently by people in different positions in these power equations.
> 
> That means, regardless of how correct the facts of a situation was, without the full extent of all the facts being known to all parties people will have different feelings on matters. That's just how humans are.
> 
> While facts do indeed not care about feelings, it is also a fact that feelings exist. :)
> 
> As such, i think it's equally important that people correcting themselves like "i didn't mean it like that, i meant xyz" be accepted; but also that people not be blamed for believing initially that words were meant as they were written.

Assuming a lack of malicious intent would have helped us here.

I understand this is a difficult period for all of us - but "don't assume on malice" is still the best policy for us to have. (And yes, myself included.)

wchristian commented:
Fwiw, malice was never assumed about this. Like, yeah, we all know this is to improve things, and we actually appreciate the improvements as well. :)

toddr commented:
Having heard no further technical concerns on this PR, I am going to merge it now. Thank you everyone for your feedback to make this merge better.

Todd
